Pushing ticket UXUI-210 for Email Component

manifestinteractive opened the pull request:
#### What does this PR do?

This is a new UI Toolkit Component.  Originally requested to be for Emails, it now supports Text, Password & Email types.

#### What unit or integration tests does this PR have?

Jest & ESLint

#### What selenium tests does this PR have?

None

#### How should a developer review this?

Peer Review then Manual Testing in TripAppLite

#### How should this be manually tested?

Test into TripAppLite

#### Any background context you want to provide?

None

#### What are the relevant tickets?

https://holidayextras.jira.com/browse/UXUI-210

#### Screenshots (if appropriate)

None

#### What gif best describes how you feel about this work?

![it_worked](http://lab.peterschmalfeldt.com/devlife/it_worked.gif)

---
#### Developer Definition of Done/Quality Checklist (for PR author to complete BEFORE code review):
- [x] I have checked the [contributing document](../blob/master/CONTRIBUTING.md) and I'm happy for this to be reviewed.

#### Software Engineer or Developer review:
- [x] I’ve witnessed the work behaving as expected (this could be on the authors machine or screencast).
- [x] I’ve checked for coding anti-patterns.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve run all the tests locally and they pass.

#### Software Engineer or project guru review:
- [ ] I’ve witnessed the work behaving as expected (this could be on the authors machine or screencast).
- [ ] I’ve checked for coding anti-patterns.
- [ ] I’ve checked for appropriate test coverage.
- [ ] I’ve run all the tests locally and they pass.

rahulpatel commented:
Taking SE/Dev review...

rahulpatel commented on src/components/input/README.md:
```diff
@@ -0,0 +1,96 @@
+# component-input
```
The component specific README files have been removed check #13.

rahulpatel commented on src/components/input/dev/example.jsx:
```diff
@@ -0,0 +1,8 @@
+var React = require('react');
```
The `dev/` folder has been removed from the other components, check PR #13.

rahulpatel commented:
Can you change tabs to spaces please?

rahulpatel commented on src/components/input/code/templates/inputTemplate.jsx:
```diff
@@ -0,0 +1,46 @@
+
+module.exports = function (props){
+
+	var classes = classNames('input-group', {
```
You're using a function called `classNames` here but it's not been defined or required in.

rahulpatel commented on src/components/input/code/templates/inputTemplate.jsx:
```diff
@@ -0,0 +1,46 @@
+module.exports = function (props){
+
+	var classes = classNames('input-group', {
+		'input-group-error': this.props.error || false,
```
Would it be better to use [`getDefaultProps`](https://facebook.github.io/react/docs/component-specs.html#getdefaultprops) for setting default values?

rahulpatel commented on src/components/input/code/templates/inputTemplate.jsx:
```diff
@@ -0,0 +1,46 @@
+				name={this.props.ref}
+				placeholder={this.props.placeholder}
+				value={this.state.value}
+				onChange={this._onChange}/>
```
Will `this._onChange` be available here or should it be `props.onChange`? I see you set `onChange` in `getDefaultProps` in `src/components/input/code/views/inputView.jsx`

rahulpatel commented on src/components/input/code/templates/inputTemplate.jsx:
```diff
@@ -0,0 +1,46 @@
+	// the form label
+	var label;
+
+	if (this.props.label){
```
Should all instances of `this.props.<key>` be `props.<key>`?

rahulpatel commented on src/components/input/__tests__/input-test.js:
```diff
@@ -0,0 +1,15 @@
+var InputComponent = require('../code/views/inputView.jsx');
+var TestUtils = React.addons.TestUtils;
+
+describe('InputComponent', function() {
```
I'm seeing a bunch of login in the template, but not many tests. Can you add tests around the logic you have within the component please?

rahulpatel replied:
I see you've added a couple tests, there is still a lot more that needs to be tested.

Here are the tests that I would expect at a minimum:
- When the component is in an error state
  - is the error class present
  - is there an error span
- When the component is disabled
  - is the disabled class present/is the element disabled
- When the component is provided with a label
  - is the label displayed

I'd also add tests for the `onChange` function testing every difference scenario.

rahulpatel commented on src/components/input/code/views/inputView.jsx:
```diff
@@ -0,0 +1,82 @@
+
+	propTypes: {
+		label: React.PropTypes.string,
+		type: React.PropTypes.string,
```
Currently I can pass any input type I want, might be worth using `React.PropTypes.oneOf([ ... ])` to lock it down to the types this component supports.

rahulpatel commented on src/components/input/code/templates/inputTemplate.jsx:
```diff
@@ -0,0 +1,46 @@
+				type={this.props.type}
+				name={this.props.ref}
+				placeholder={this.props.placeholder}
+				value={this.state.value}
```
`this.state` is not in scope, same goes for `this.props`.

rahulpatel commented:
I think i've reviewed all that I can, i'll review it once all comments have been addressed.

manifestinteractive commented:
Update component to greatly simplify validating inputs.  Also addressed issues raised in PR comments.  Ready for another review.

rahulpatel commented on src/components/input/code/templates/inputTemplate.jsx:
```diff
@@ -0,0 +1,48 @@
+        name={component.props.name}
+        id={component.props.id}
+        placeholder={component.props.placeholder}
+        onChange={component.changeHandler}
```
You're using the `changeHandler` function that you've defined within the component directly here, is it supposed to be the `onChange` function passed in via props?

You define `onChange` within the `propTypes` object and assign `this.changeHandler` to `onChange` in `getDefaultProps`.

rahulpatel replied:
Hi @manifestinteractive,

Could you please address this comment?

Thanks,
Rahul

rahulpatel commented on src/components/input/code/views/inputView.jsx:
```diff
@@ -0,0 +1,61 @@
+      is_valid = this.props.validator.test(e.target.value);
+    }
+
+    if( !is_valid && typeof this.props.errorMessage !== 'undefined'){
```
You can get rid of the `else` part of this if statement by setting `this.props.errorMessage` to `null` in `getDefaultProps`.

rahulpatel commented:
There are still a few things that need to addressed, other than that it's looking better :).

The other components have branded styling attached to them, i'm wondering if this component needs the same thing, @jodiedoubleday can I get your _input_ here please?

jodiedoubleday commented:
Nice punn @rahulpatel at the moment I'm happy for this to have default input styling. Going to be working more closelt with @stuartgiles to create a better standard of styling

jodiedoubleday commented:
Maybe not for this PR, but we should make sure it uses HTML5 validation by default, and then 
fall back

jodiedoubleday commented:
we should have a `for="foo"` on the label which links to the `id="foo"` on the input

manifestinteractive commented:
__Addressed the following:__

1.  Merged in master and addressed merge conflicts
2.  Added for to label ( this was in there before, but was listed as for="" rather than reacts htmlFor=""
3.  Updated Unit Tests with apparent new way of doing unit test :smirk: 
4.  Added new Unit Tests to make sure label, disabled & readonly states are checked
5.  npm test && npm run lint pass

__Blocking:__

There is a request to create a unit test for the onChange event, which would require creating the component and then somehow triggering a state change on it.  I am not sure how to do this within our test suite, anyone got an idea?

hxtomprice commented:
Taking IP/Dev

hxtomprice commented:
When I run this locally with `grunt docs` in chrome the validation doesn't clear for me: http://grab.by/HMhs 
It also validates as soon when key is pressed but reading the comments in the jira it sounds like there should be a delay before validation fires. I'm not sure if this is a feature which is no longer wanted

manifestinteractive commented:
This is ready review. Added delay to not fire validation until keypress briefly stops.  Also updated to clear validation.

hxtomprice commented:
Looks great on desktop, when I shrink the page down to mobil size the labels start to get lost under the input fields: http://grab.by/HOBI

manifestinteractive commented:
@hxtomprice please speak with @jackdcrawford regarding the style related issues.  Currently the UXUI is working on standardizing where the styles will be stored ( either externally in a theme, or internally as a  LESS file ).  The current styling of this component is for the Docs only, and was not intended as the permanent style ( assumption being the UXUI team would release an official theme ).

hxtomprice commented:
Happy to +1 this as the styles are going to be sorted later

rahulpatel commented on src/components/input/code/views/inputView.jsx:
```diff
@@ -0,0 +1,68 @@
+    };
+  },
+
+  handleChange: function(e) {
```
Could you add a test to ensure this is working as expected?

I'd like to see a test where the component **isn't** passed an `errorMessage` and a test where the component **is** passed an `errorMessage`. Both of these tests should trigger an error and check the message rendered is what's expected.

manifestinteractive commented:
@rahulpatel Added requested unit tests and merged in latest master.  Let's see if we can get this PR through today, that would make everyones day, right? ( The "Merge pull request" button looks so pretty today :smile_cat: ).

rahulpatel commented on src/components/input/__tests__/input-test.js:
```diff
@@ -0,0 +1,105 @@
+
+  it('should change the value to be invalid and throw error state with null message', function() {
+    TestUtils.Simulate.change(renderedInput.getDOMNode(), { target: { value: 'invalid @ email . #$%' }});
+    assert.equal(input.state.unitTestError, null);
```
Do we want to set a default error message? It'll be better than just showing nothing/null to the user?

rahulpatel commented:
Sorry I couldn't pick this up yesterday @manifestinteractive, only 1 minor comment, otherwise this looks ready to go! :dancer:

manifestinteractive commented:
@rahulpatel Your last remaining issue was addressed, I also added more tests based on feedback I got from @lukehansell-hx on other components.  Latest master is also merged into this commit.

rahulpatel commented:
Happy with this :+1:!
